### PR TITLE
client, mds: report and validate client's mount root

### PR DESCRIPTION
**Pull request by jcsp**

### CODE REVIEW

**Comment by jcsp** [comment]
I've expanded this quite a bit to add a way for MDSs to close a client's session and also tell the client not to bother us again.  We might want this in future for any other situations where a client is identified as badly behaved?  Since it's a protocol extension, ping: @liewegas @gregsfortytwo

File: src/mds/MDSAuthCaps.cc [hunk 1]
```diff
@@ -126,10 +136,26 @@ bool MDSCapMatch::match(const std::string &target_path,
...
   }
+
   return true;
 }
 
 /**
+ * Is the client *potentially* able to access this path?  Actual
+ * permission will depend on uids/modes in the full is_capable.
+ */
+bool MDSAuthCaps::maybe_capable(const std::string &inode_path) const
```
Comment (gregsfortytwo) [inline src/mds/MDSAuthCaps.cc]: How about we call it path_capable instead of maybe_capable.

**Comment by gregsfortytwo** [comment]
Do you envision REJECTing a client for anything other than stuff that looks like attempts to hack the system? I don't, and so I think we should blacklist rejected clients.

And clients which get a REJECT should just politely shut down with a terrifying error message.

**Comment by jcsp** [comment]
@gregsfortytwo the blacklisting and shutting down points are kind of the same thing: making reject a permanent thing rather than a "go wait" thing.  The reason I made reject a non-deadly behaviour was that in future we might change this on the MDS side, and if a client sees a reject from one MDS, that doesn't necessarily mean it should die, it should wait til it sees an MDS that is willing to talk to it.

As for things other than hacking the system, I can imagine using this mechanism to impose client-restricting soft policies, like only permitting above a certain version of clients.  We are permissive by default, but I can imagine an administrator or a support person wanting to configure a system that rejected clients below a certain version due to known bugs.

File: src/mds/MDSAuthCaps.cc (hunk 1 above)
Comment (jcsp) [inline src/mds/MDSAuthCaps.cc]: Sure, why not (done)

**Comment by gregsfortytwo** [comment]
Mmm. There's no way we'll be running a mixed MDS cluster where some of them are willing to talk to a client but others aren't, and "reject" is the right answer — I can't even imagine maintaining a filesystem structure like that.
Things like trying to restrict clients to a subset of the servers (for geographically distributed systems or whatever) need very different semantics from "I don't believe you". Even if we do want soft policies about client versions, we still want them to die, not inexplicably sit idly. I'd much prefer well-behaved clients to die when they see this; it's easy to understand and makes sense.

**Comment by jcsp** [comment]
This has a new test: https://github.com/ceph/ceph-qa-suite/pull/743

**Comment by jcsp** [comment]
Rebased.  Can't remember if I pointed this out out of band last year, but the latest code has the behaviour of clients dying on mount when they're rejected (that's what the test exercises).

**Comment by gregsfortytwo** [comment]
Between http://pulpito.ceph.com/gregf-2016-01-12_23:23:33-fs-greg-fs-testing-1-12-1---basic-mira/ and http://pulpito.ceph.com/gregf-2016-01-13_16:11:43-fs-greg-fs-testing-1-12-2---basic-mira/ I think we're in good shape here.

**Pull request merged**